tendermint/src/amino_types/block_id.rs

```diff
@@ -34,7 +34,7 @@ impl block::ParseId for BlockId {
 
 impl From<&block::Id> for BlockId {
     fn from(bid: &block::Id) -> Self {
-        let bid_hash = bid.hash.as_bytes().unwrap().to_vec();
+        let bid_hash = bid.hash.as_bytes().unwrap_or(&[]).to_vec();
```

Comment:
This seems like a weird way of addressing the problem... doesn't it result in an invalid BlockId?

I'd suggest changing this whole thing to either:

```rust
impl From<&block::Id> for BlockId
```

or

```rust
impl TryFrom<&block::Id> for BlockId
```

Reply:
This is how golang tendermint handled this. I agree we should do better. But we have to handle the serialization identical to golang version to compute signing bytes, what do you think of that?

Reply:
In Rust I think it's more idiomatic to reflect the distinction using the type system

Reply:
I think we can also remove the Option in the ``parts_header`` field. I think when ``hash`` is not empty, ``parts_header`` can't be empty.